### DISCUSSION TITLE
Problem with running .net aspire project on linux

### DISCUSSION THREAD

**Discussion by wojteksuwala** [microsoft/aspire#1381] [General]
I generated a basic .net aspire project on Ubuntu. When I run it using dotnet run I get web api project running on the port specified in launchSettions.json (5514) but when I inspect Blazor app env variables using dashboard I get completely different ports - for example http://localhost:44029. What am I doing wrong?

**Comment by davidfowl** [comment]
Are you talking about the URL to talk to the backend or something else?

**Reply by wojteksuwala** [reply]
Exactly. When AppHost starts, I go to the dashboard. From there I can access both web API and client at ports defined in launchsettings but the client app is unable to get data from web API, as it tries to call it on a different port.
I tried to debug builder.Build().Run(); and get to ApplicationExecutor::RunApplicationAsync and found so far:
- before a call to CreateServicesAsync in PrepareServices I see that metadata for service definition contains ServiceBindingAnnotation which points to host/port from launchSettings, but the created instance of Service does not have AllocatedPort set
- call to CreateResourcesAsync results in srvResource.Service.ApplyAddressInfoFrom(updated); in a foreach loop after kubernetesService.WatchAsync and address of web API is set but not to address from launchSettings.json

**Comment by wojteksuwala** [comment]
It may be worth mentioning that I am not running Docker Desktop with Kubernetes support but k3d local cluster.

**Comment by wojteksuwala** [comment]
Here are the screenshots from the dashboard

![services](https://github.com/dotnet/aspire/assets/5447214/74637393-c008-4496-9337-e5904b84be4c)

![env_vars](https://github.com/dotnet/aspire/assets/5447214/d2d3d23d-20fb-4d03-9182-2fe2fb8650e6)

**Comment by davidfowl** [comment]
The orchestrator produces proxies that use a different URL. Those proxies then route to the backend service. We're cleaning it up in preview 2 so that the URLs are as expected.

**Comment by wojteksuwala** [comment]
Big thanks for your help. In my case the proxy doesn't work, could you direct me to documentation or place in code that is responsible for it or describe how it works?

**Comment by wojteksuwala** [comment]
From logs I can see that the proxy process keeps restarting - how can I find what is causing it?

**Reply by davidfowl** [reply]
Can you share the logs?

**Reply by wojteksuwala** [reply]
nfo: Aspire.Dashboard.DashboardWebApplication[0]
      Now listening on: http://localhost:15061
info: Aspire.Dashboard.DashboardWebApplication[0]
      OTLP server running at: http://localhost:16213
info: Aspire.Hosting.Dcp.DcpHostService[0]
      Starting DCP with arguments: start-apiserver --monitor 24018 --detach --kubeconfig "/tmp/aspire.34dENl/kubeconfig"
info: Aspire.Hosting.Dcp.start-apiserver.dcp-host[0]
      Starting DCP API server
info: Aspire.Hosting.Dcp.start-apiserver.dcp-host[0]
      Starting DCP controller host
info: Aspire.Hosting.Dcp.start-apiserver.dcp-host[0]
      Started all services      {"count": 2}
info: Aspire.Hosting.Dcp.api-server[0]
      Starting API server...
info: Aspire.Hosting.Dcp.api-server[0]
      API server started        {"Address": "127.0.0.1", "Port": 33201}
info: Aspire.Hosting.Dcp.dcpctrl.IdeExecutableRunner[0]
      Executables cannot be started via IDE: missing required environment variable 'DEBUG_SESSION_PORT'
info: Aspire.Hosting.Dcp.dcpctrl[0]
      starting controller manager
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24095 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 3}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      service /apiservice is now in state NotReady      {"ServiceName": {"name":"apiservice"}, "Reconciliation": 3}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24096 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 4}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      service /webfrontend is now in state NotReady     {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 4}
info: Aspire.Hosting.Dcp.dcpctrl.ExecutableReconciler[0]
      starting process...       {"Controller": "usvc-dev.developer.microsoft.com/executable-reconciler", "Executable": {"name":"apiservice"}, "Reconciliation": 2, "State": "", "ExecutionID": "", "executable": "/usr/bin/dotnet"}
info: Aspire.Hosting.Dcp.dcpctrl.ExecutableReconciler[0]
      process started   {"Controller": "usvc-dev.developer.microsoft.com/executable-reconciler", "Executable": {"name":"apiservice"}, "Reconciliation": 2, "State": "", "ExecutionID": "", "executable": "/usr/bin/dotnet", "PID": 24097}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      service /apiservice is now in state Ready {"ServiceName": {"name":"apiservice"}, "Reconciliation": 7}
info: Aspire.Hosting.Dcp.dcpctrl.ExecutableReconciler[0]
      Executable run changed    {"Controller": "usvc-dev.developer.microsoft.com/executable-reconciler", "Executable": {"name":"apiservice"}, "Reconciliation": 3, "State": "", "ExecutionID": "", "RunID": "24097", "PID": null, "NewPID": 24097, "State": "", "NewState": "Running", "ExitCode": null, "NewExitCode": null}
info: Aspire.Hosting.Dcp.dcpctrl.ExecutableReconciler[0]
      starting process...       {"Controller": "usvc-dev.developer.microsoft.com/executable-reconciler", "Executable": {"name":"webfrontend"}, "Reconciliation": 6, "State": "", "ExecutionID": "", "executable": "/usr/bin/dotnet"}
info: Aspire.Hosting.Dcp.dcpctrl.ExecutableReconciler[0]
      process started   {"Controller": "usvc-dev.developer.microsoft.com/executable-reconciler", "Executable": {"name":"webfrontend"}, "Reconciliation": 6, "State": "", "ExecutionID": "", "executable": "/usr/bin/dotnet", "PID": 24100}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      service /webfrontend is now in state Ready        {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 9}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /home/wsuwala@software.altkom.pl/Projects/gildia/dotnet8/aspire/aspire.AppHost
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24201 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 11}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24205 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 13}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24320 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 15}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24332 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 17}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24353 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 19}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24363 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 21}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24385 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 23}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24395 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 25}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24407 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 27}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24417 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 29}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24428 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 31}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24440 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 33}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24454 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 35}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24464 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 37}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24480 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 39}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24489 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 41}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24501 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 43}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24512 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 45}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24523 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 47}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24533 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 49}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24545 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 51}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24556 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 53}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24566 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 55}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24573 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 57}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24587 started for service /apiservice
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /apiservice    {"ServiceName": {"name":"apiservice"}, "Reconciliation": 59}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process with PID 24594 started for service /webfrontend
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process has been started for service /webfrontend   {"ServiceName": {"name":"webfrontend"}, "Reconciliation": 61}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /webfrontend exited with code 1
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      proxy process for service /apiservice exited with code 1
^Cinfo: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
info: Aspire.Hosting.Dcp.dcpctrl.ExecutableReconciler[0]
      Executable is being deleted...    {"Controller": "usvc-dev.developer.microsoft.com/executable-reconciler", "Executable": {"name":"apiservice"}, "Reconciliation": 10, "State": "Running", "ExecutionID": "24097"}
info: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler[0]
      Service object is being deleted   {"ServiceName": {"name":"apiservice"}, "Reconciliation": 63}

**Comment by wojteksuwala** [comment]
After upgrading to preview 2 the urls look as expected but proxies seem to be restarting again and again and I cannot connect to frontend or API using URLs provided in the dashboard. Is there a way to log more info/debug this issue?

**Reply by davidfowl** [reply]
cc @karolz-ms @dbreshears

**Comment by karolz-ms** [comment]
@wojteksuwala thanks for the report. You can get more detailed logs by setting the `DCP_DIAGNOSTICS_LOG_LEVEL` environment variable to `debug`. The logs will be put to `${TEMP}/dcp/logs`, and you can change the destination by setting `DCP_DIAGNOSTICS_LOG_FOLDER` environment variable.

Is there a chance to share your solution with us by any chance?

**Reply by wojteksuwala** [reply]
Thanks for your answer. The solution is generated from the template, built, and run without any changes. I'll try later to run the app with env vars set and if I find anything usefull I'll share here.
